### ISSUE TITLE
RFC: dynamically-typed runtime

### ISSUE THREAD

**Issue by mihaibudiu** [feldera/dbsp#330]
There's an alternative design for the runtime: using dynamically-typed union types.
(This would be somewhat difficult for the JIT, though, at least with the current design. The JIT is strongly-typed, but you could add a union type there as well.)

One nice thing about such a design is that one of the value types can be Error. This would allow building panic-free pipelines, and propagating errors through the computation - you could even keep the lineage of an error output record. This would also make unnecessary the existence of an error-reporting interface between the data-plane and the control-plane.
There are downsides too: more runtime overhead, no strong type-checking, wrappers are needed for all library functions, but maybe some of these can be generated using macros.

**Comment by mihaibudiu** [comment]
Certainly, we could also keep the stronly-typed design and wrap every type we have in a Result, but that would be ugly.

**Comment by ryzhyk** [comment]
Propagating errors through the pipeline seems very infectious, e.g., I imagine that aggregating  a set of records one of which is an `Error` produces an `Error`, so user will all of a sudden observe cascades or `Error` outputs, which they have no good way of processing.

**Comment by ryzhyk** [comment]
> Certainly, we could also keep the stronly-typed design and wrap every type we have in a Result, but that would be ugly.

I like the Rust approach, with the `?` operator, which allows returning an error from any part of a function, so as long as DBSP allows all closures to return errors, we're good.  But I guess `?` can be tricky to implement in JIT, as it requires deallocating any live objects in the current scope.

**Comment by mihaibudiu** [comment]
Infectious may be better than dead.
